### PR TITLE
creature: initialise room number for target enemy

### LINKED ISSUE #137: TR2X bug: Rare crash in Barkhang Monastery
**Dev snapshot:** 0.2-227-g6e863b2

__________________

The cause or steps to reproduce this crash is TBD. Hopefully this issue will provide a clue to help figure out a fix.

The waypoints shown below mark the 3 places I experienced a crash.  In all cases, it was after Lara collected a Prayer Wheel in room 45.

Screenshot

### PR DESCRIPTION
**Pull request by lahm86**
Resolves #137.

This avoids cases with enemies fighting one another where x and z positions are identical and so `LOS_Check` does not initialise the target room number, and hence later passes an invalid room number to `Room_GetSector`.

I got the trace to show it was stemming from the call in `Creature_CanTargetEnemy`. I realise the room number isn't being set elsewhere for other LOS check targets, but felt perhaps this should be kept in its own commit for now while we continue testing the rest.

I tested the Barkhang scenario for an hour (i.e. loading a save, waiting for monk/merc fights to play out, then reloading) and it was crash-free. @aredfan, perhaps you could also check it out if you have time?
[TR2X.zip](https://github.com/user-attachments/files/16709274/TR2X.zip)

### CODE REVIEW

**Comment by aredfan** [comment]
Thank you for working on this PR @lahm86.

I've not experienced any crashes while testing this build. I did notice an issue with enemies being able to target and shoot Lara/monks through walls.

**Saves:**
[savegame.zip](https://github.com/user-attachments/files/16710523/savegame.zip)

**Comment by lahm86** [comment]
> Thank you for working on this PR @lahm86.
> 
> I've not experienced any crashes while testing this build. I did notice an issue with enemies being able to target and shoot Lara/monks through walls.
> 
> **Saves:** [savegame.zip](https://github.com/user-attachments/files/16710523/savegame.zip)

Thank you. Back to the drawing board 😄

**Comment by lahm86** [comment]
The issue with enemies shooting through walls is now resolved (thank you @rr-), this was a problem only in the release build which I had sent. Would you mind checking over it again @aredfan? So this is the same fix as before for the original crashing problem, plus the shooting issue resolved as well.
[TR2X.zip](https://github.com/user-attachments/files/16716157/TR2X.zip)

File: src/game/creature.c [hunk 1]
```diff
@@ -913,6 +913,7 @@ int32_t __cdecl Creature_CanTargetEnemy(
     target.pos.x = enemy->pos.x;
     target.pos.y = enemy->pos.y - STEP_L * 3;
     target.pos.z = enemy->pos.z;
+    target.room_num = enemy->room_num;
```
Comment (rr-) [inline src/game/creature.c:916]: I'm a bit torn here; can we use the previous fix where we always set the target room number, and on top of it, have every caller set the room number as well?

This way we make both the callers and the LOS_Check API behave nicely – the latter will make sure the room number is always set to a valid value, and the former will always provide a nice input to the function. Ideally we could use this same approach in TR1X.

(I'm open to discussion about this.)

Comment (rr-) [reply src/game/creature.c:916]: Another idea is to alter the LOS internals to use `XYZ_32` for `target`, and have `LOS_Check` set the `target->room_num` to `Room_FindByPos` result.

Comment (lahm86) [reply src/game/creature.c:916]: Yes, I'm happy to put the check back into `LOS_Check`, then have each caller explicitly set it. It feels right for sure.
I think we may need to have `Gun_FindTargetPoint` decompiled as well first though, as @walkawayy mentioned:
https://github.com/LostArtefacts/TR2X/issues/137#issuecomment-2303347091

**Comment by aredfan** [comment]
> The issue with enemies shooting through walls is now resolved (thank you @rr-), this was a problem only in the release build which I had sent. Would you mind checking over it again @aredfan? So this is the same fix as before for the original crashing problem, plus the shooting issue resolved as well. [TR2X.zip](https://github.com/user-attachments/files/16716157/TR2X.zip)

Of course, no problem. I did 30 minutes of tests with the newer build and there's been no crashes, or any other odd issues.

I'll be happy to do more tests tomorrow if the code changes. 👍

File: src/game/creature.c (hunk 1 above)
Comment (rr-) [reply src/game/creature.c:916]: Done

Comment (walkawayy) [reply src/game/creature.c:916]: Thanks! Looks like `Gun_FindTargetPoint` does set `target->room_num = item->room_num;`, so that situation should be good.

Comment (lahm86) [reply src/game/creature.c:916]: Pushed a change to have all calls set the room number first, and `LOS_Check` also does a check for identical x/z. I can squash when the time comes as the title/commits have changed a bit already for this one.

[TR2X.zip](https://github.com/user-attachments/files/16717945/TR2X.zip)

The other main references to the function are in `camera.c` but everything here checks out already as far as I can tell.

**Review by walkawayy — APPROVED** [review]
LGTM. I didn't see any regressions or crashes. I killed a monk and kept a locked on target on his dead body across the level to make sure shooting through rooms / target lock remained. Did we figure out if this and the issue should be tagged OG bug or TR2X bug?

**Review by rr- — APPROVED** [review]

**Review by aredfan — APPROVED** [review]
LGTM.

**Pull request merged**